src/imageops/mod.rs

```diff
@@ -81,72 +79,65 @@ fn crop_dimms<I: GenericImageView>(
 /// * Overflows in the computation
 /// * Coordinates could be completely out of bounds
 ///
-/// The main idea is to make use of inequalities provided by the nature of `saturing_add` and
-/// `saturating_sub`. These intrinsically validate that all resulting coordinates will be in bounds
-/// for both images.
-///
-/// We want that all these coordinate accesses are safe:
-/// 1. `bottom.get_pixel(x + [0..x_range), y + [0..y_range))`
-/// 2. `top.get_pixel([0..x_range), [0..y_range))`
-///
-/// Proof that the function provides the necessary bounds for width. Note that all unaugmented math
-/// operations are to be read in standard arithmetic, not integer arithmetic. Since no direct
-/// integer arithmetic occurs in the implementation, this is unambiguous.
-///
-/// ```text
-/// Three short notes/lemmata:
-/// - Iff `(a - b) <= 0` then `a.saturating_sub(b) = 0`
-/// - Iff `(a - b) >= 0` then `a.saturating_sub(b) = a - b`
-/// - If  `a <= c` then `a.saturating_sub(b) <= c.saturating_sub(b)`
-///
-/// 1.1 We show that if `bottom_width <= x`, then `x_range = 0` therefore `x + [0..x_range)` is empty.
-///
-/// x_range
-///  = (top_width.saturating_add(x).min(bottom_width)).saturating_sub(x)
-/// <= bottom_width.saturating_sub(x)
-///
-/// bottom_width <= x
-/// <==> bottom_width - x <= 0
-/// <==> bottom_width.saturating_sub(x) = 0
-///  ==> x_range <= 0
-///  ==> x_range  = 0
-///
-/// 1.2 If `x < bottom_width` then `x + x_range < bottom_width`
+/// The returned value is of the form:
 ///
-/// x + x_range
-/// <= x + bottom_width.saturating_sub(x)
-///  = x + (bottom_width - x)
-///  = bottom_width
+/// `(origin_bottom_x, origin_bottom_y, origin_top_x, origin_top_y, x_range, y_range)`
 ///
-/// 2. We show that `x_range <= top_width`
+/// The main idea is to do computations on i64's and then clamp to image dimensions.
+/// In particular, we want to ensure that all these coordinate accesses are safe:
+/// 1. `bottom.get_pixel(origin_bottom_x + [0..x_range), origin_bottom_y + [0..y_range))`
+/// 2. `top.get_pixel(origin_top_y + [0..x_range), origin_top_y + [0..y_range))`
 ///
-/// x_range
-///  = (top_width.saturating_add(x).min(bottom_width)).saturating_sub(x)
-/// <= top_width.saturating_add(x).saturating_sub(x)
-/// <= (top_wdith + x).saturating_sub(x)
-///  = top_width (due to `top_width >= 0` and `x >= 0`)
-/// ```
-///
-/// Proof is the same for height.
 pub fn overlay_bounds(
     (bottom_width, bottom_height): (u32, u32),
     (top_width, top_height): (u32, u32),
-    x: u32,
-    y: u32
-)
-    -> (u32, u32)
-{
-    let x_range = top_width.saturating_add(x) // Calculate max coordinate
-        .min(bottom_width) // Restrict to lower width
-        .saturating_sub(x); // Determinate length from start `x`
-    let y_range = top_height.saturating_add(y)
-        .min(bottom_height)
-        .saturating_sub(y);
-    (x_range, y_range)
+    x: i64,
+    y: i64,
+) -> (u32, u32, u32, u32, u32, u32) {
```

Comment:
That's decently complex. How about having a simple struct for this with public members and methods to construct the x-y-ranges (`core::ops::Range<u32>`) for the bot and top image. I wouldn't mind as much if it weren't part of the public API.

Reply:
How important is it to keep the improved overlay_bounds function as part of the public API?

Reply:
Ok, this PR now leaves `overlay_bounds` unchanged and converts internal uses to a private overlay_bounds_ext function instead